### PR TITLE
Add PonyCheck to the tutorial

### PR DESCRIPTION
**Pull request by SeanTAllen**

### CODE REVIEW

**Comment by SeanTAllen** [comment]
Per the RFC, this content is repurposed from the PonyCheck README.

I'm trying to find a couple of good "introduction to property based testing" resources to include as well with the target audience of "people who know unit-testing and either don't know what property-based testing is or have heard of it but have never done it".

**Comment by SeanTAllen** [comment]
This will fail checks until the Ponyc PR is merged.

**Comment by SeanTAllen** [comment]
Shouldn't be merged until after pony 0.49.0 has been released.

File: docs/testing/ponycheck.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ PonyCheck is heavily inspired by QuickCheck and other great property based testi
 Writing property based tests in PonyCheck is done by implementing the trait [`Property1`](https://stdlib.ponylang.io/PonyCheck-Property1). A [`Property1`](https://stdlib.ponylang.io/PonyCheck-Property1) needs to define a type parameter for the type of the input sample, a [`Generator`](https://stdlib.ponylang.io/PonyCheck-Generator) and a property function. Here is a minimal example:
```
Comment (rhagenson) [inline docs/testing/ponycheck.md:19]:
```suggestion
Writing property based tests in PonyCheck is done by implementing the trait [`Property1`](https://stdlib.ponylang.io/pony_check-Property1). A [`Property1`](https://stdlib.ponylang.io/pony_check-Property1) needs to define a type parameter for the type of the input sample, a [`Generator`](https://stdlib.ponylang.io/pony_check-Generator) and a property function. Here is a minimal example:
```

File: docs/testing/ponycheck.md [hunk 2]
```diff
@@ -61,11 +61,11 @@ class _ListReverseOneProperty is Property1[Array[USize]]
 
 ## Integration with PonyTest
 
-PonyCheck properties need to be executed. The test runner for PonyCheck is [PonyTest](https://stdlib.ponylang.org/ponytest--index). To integrate [`Property1`](https://stdlib.ponylang.io/PonyCheck-Property1) into [PonyTest](https://stdlib.ponylang.org/ponytest--index), `Property1` needs to be wrapped inside a [`Property1UnitTest`](hhttps://stdlib.ponylang.io/PonyCheck-Property1UnitTest) and passed to the PonyTest `apply` method as all regular PonyTest [`UnitTests`](https://stdlib.ponylang.org/ponytest-UnitTest):
+PonyCheck properties need to be executed. The test runner for PonyCheck is [PonyTest](https://stdlib.ponylang.org/pony_test--index). To integrate [`Property1`](https://stdlib.ponylang.io/PonyCheck-Property1) into [PonyTest](https://stdlib.ponylang.org/pony_test--index), `Property1` needs to be wrapped inside a [`Property1UnitTest`](hhttps://stdlib.ponylang.io/PonyCheck-Property1UnitTest) and passed to the PonyTest `apply` method as all regular PonyTest [`UnitTests`](https://stdlib.ponylang.org/pony_test-UnitTest):
```
Comment (rhagenson) [inline docs/testing/ponycheck.md:64]:
```suggestion
PonyCheck properties need to be executed. The test runner for PonyCheck is [PonyTest](https://stdlib.ponylang.org/pony_test--index). To integrate [`Property1`](https://stdlib.ponylang.io/pony_check-Property1) into [PonyTest](https://stdlib.ponylang.org/pony_test--index), `Property1` needs to be wrapped inside a [`Property1UnitTest`](hhttps://stdlib.ponylang.io/pony_check-Property1UnitTest) and passed to the PonyTest `apply` method as all regular PonyTest [`UnitTests`](https://stdlib.ponylang.org/pony_test-UnitTest):
```

**Review by rhagenson — CHANGES_REQUESTED** [review]
I took a look at the raw file in addition to the diff. Raw file changes outside the diff are commented about below:

+ "Property based Testing first came up as..." -> "Property based testing first came up as..." i.e. lowercase "t"
+ "A Property needs a name" -> "A Property1 needs a name" ie add the "1"
+ "Below is a classical List reverse properties from the QuickCheck paper adapted to Pony Arrays" -> "Below is a classical `List` reverse properties from the QuickCheck paper adapted to Pony `Array`s" ie add the code formatting for referenced data structures (or lowercase them so it is simply list and arrays which, to me, would then be generic and need no special formatting)
+ Many of the links are/will be broken right now, referring to addresses such as https://stdlib.ponylang.io/PonyCheck-Property1 rather than https://stdlib.ponylang.io/pony_check-Property1 ie correctly adding the underscore in "pony_check" per the rename

**Comment by SeanTAllen** [comment]
@rhagenson can you do suggested changes for fixes or go ahead and edit directly and push?

**Comment by rhagenson** [comment]
I can, yes. Give me a couple minutes to hop back on my laptop and the changes will be in place.

**Comment by SeanTAllen** [comment]
I made what I could figure out.

**Comment by rhagenson** [comment]
I will take a look now and push any remaining changes.

File: docs/testing/ponycheck.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    PonyCheck.forAll[Array[USize]](gen1, h)({
+      (arg1: Array[USize], ph: PropertyHelper) =>
+        ph.assert_array_eq[USize](arg1, arg1.reverse().reverse())
+    })
+    let gen2 = Generators.seq_of[USize, Array[USize]](1, Generators.usize())
+    PonyCheck.forAll[Array[USize]](gen2, h)({
+      (arg1: Array[USize], ph: PropertyHelper) =>
+        ph.assert_array_eq[USize](arg1, arg1.reverse())
+    })
+```
```
Comment (rhagenson) [inline docs/testing/ponycheck.md:96]: Is it supposed to be `PonyCheck.for_all` as in the current API of ponylang/ponycheck or has this been changed to `PonyCheck.forAll` per style guide changes?

**Comment by rhagenson** [comment]
@SeanTAllen My edits are in place now. You fixed much of what I had suggested as fixes already, but I found some inconsistencies in the URL used for the stdlib (.io versus .org), made a consistency edit to name `PropertyHelper` parameters the same across this page (as `ph` rather than `h`), and modified the language to avoid having to say "UnitTests" then linking to a page for singular "UnitTest" page.

**Review by rhagenson — APPROVED** [review]

File: docs/testing/ponycheck.md (hunk 3 above)
Comment (SeanTAllen) [reply docs/testing/ponycheck.md:96]: I didn't that issue. Can you update here and update the PR on ponyc including noting the change in the release notes?

Comment (SeanTAllen) [reply docs/testing/ponycheck.md:96]: So snake case is used for functions so forAll would be incorrect.

Comment (SeanTAllen) [reply docs/testing/ponycheck.md:96]: It looks like that is only in the tutorial here, so I will update here.

Comment (rhagenson) [reply docs/testing/ponycheck.md]: Perfect! Thank you!

**Pull request merged**